### PR TITLE
Should have specific ID JSON rule

### PR DESCRIPTION
**Pull request by UnknownNPC**
PR is related to the issue #268.
There are some code with `SpecificJsonRule`(https://github.com/mafagafogigante/dungeon/issues/268#issuecomment-289222028), tests for it and some usage samples inside `ItemsJsonFileTest` and `CreaturesJsonFileTest` test files. 

I should note, all constants with JSON file names were replaced to `ResourcesTypeTest`, because some tests require additional files for pass (eg. `ItemsJsonFileTest` uses `locations.json` and `creatures.json`). 

Also, there are `List` constants for `	exclusion` items. I didn't expect that they will be so long.
Let's review it together to make it better.

### CODE REVIEW

File: src/main/java/org/mafagafogigante/dungeon/schema/rules/JsonRuleFactory.java [hunk 1]
```diff
@@ -75,4 +76,8 @@ public static JsonRule makePeriodRule() {
     return new PeriodJsonRule();
   }
 
+  public static JsonRule makeSpecificIdJsonRule(Collection<String> ids) {
```
Comment (mafagafogigante) [inline src/main/java/org/mafagafogigante/dungeon/schema/rules/JsonRuleFactory.java]: Why not a collection of Id objects? These would certainly be valid, these strings may be anything.

File: src/main/java/org/mafagafogigante/dungeon/schema/rules/SpecificIdJsonRule.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+
+class SpecificIdJsonRule extends IdJsonRule {
+
+  private final Set<Id> ids = new HashSet<>();
+
+  SpecificIdJsonRule(Collection<String> stringIds) {
+    for (String stringId : stringIds) {
```
Comment (mafagafogigante) [inline src/main/java/org/mafagafogigante/dungeon/schema/rules/SpecificIdJsonRule.java]: If this gets changed to Collection<Id>, prefer using addAll(), even if the JVM does not perform any optimization, at least it is shorter and easier to read.

File: src/main/java/org/mafagafogigante/dungeon/schema/rules/SpecificIdJsonRule.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+      ids.add(new Id(stringId));
+    }
+  }
+
+  @Override
+  public void validate(JsonValue value) {
+    super.validate(value);
+    Id testId = new Id(value.asString());
+    boolean isValueExist = ids.contains(testId);
```
Comment (mafagafogigante) [inline src/main/java/org/mafagafogigante/dungeon/schema/rules/SpecificIdJsonRule.java]: `exists` would be a more grammatically correct name. I don't have any problem understanding you, but the codebase should stay as readable as possible.

File: src/main/java/org/mafagafogigante/dungeon/schema/rules/SpecificIdJsonRule.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    if (!isValueExist) {
+      throw new IllegalArgumentException(testId + " is not exist in next rule id list: " + ids + ".");
```
Comment (mafagafogigante) [inline src/main/java/org/mafagafogigante/dungeon/schema/rules/SpecificIdJsonRule.java]: Change it to `testId + " is not a valid Id"`

Also, as per #347 I decided to add as a convention that exception messages do not end in punctuation (they are not sentences).

**Comment by UnknownNPC** [comment]
@mafagafogigante, thank you for your review. I have added changes required. 

Also, I revised `List` constants, it should be more clear now.  
In addition, our changes сaught unused `PAGE_FROM_VOLUND_LOKE_FREY_S_DIARY` item id inside `items.json`. But I think it's another story. : )

**Comment by mafagafogigante** [comment]
@UnknownNPC thanks for the heads up. I don't recall why it wasn't being placed. I created #370 and will work on it soon enough.

**Comment by mafagafogigante** [comment]
@UnknownNPC I gave myself the freedom to push on top of your commits to speed things up.

I was reviewing your work and noticed that we had a communication problem. You wrote the ID tests in the wrong direction.

All creature IDs that are used (in location spawners) must be declared.
But not all declared creatures must be used (in the review I changed the specification, because we may want to keep some old definitions if they ever need to be available in a saved world).

So you should build a set of all the IDs from the creatures JSON and check that every creature ID you find in the locations JSON is a valid creature ID, not the other way around.

By the looks of it, after fixing this part (I already reverted the specific rule on the creatures tests, you have to change `getSpawnersRule()` in the LocationsJsonFileTest) and changing that test inheritance into an enum (if possible) we should be good to go.

If you want to leave the parent-to-enum rewrite to me, it's fine. If you have any questions, ask them.

**Comment by mafagafogigante** [comment]
@UnknownNPC I will review this in about 13~14 hours from now, I have to go out now, but it seems we understood each other at the end.

**Comment by UnknownNPC** [comment]
@mafagafogigante, Sure. I assumed ID-relation-issues from my side, thank you for review and clarification.

**Comment by mafagafogigante** [comment]
We missed checking that the creature inventories where valid item IDs.

You apparently did not remove the check for items being used, which we don't need. I finished these parts.

I will give the PR a full reread at the end of tomorrow (April 7) and merge it if I don't find anything that needs to change.

Thank you for your contribution.

**Comment by mafagafogigante** [comment]
@UnknownNPC finally got back to it! Got this into the branch seven today, closing because I rebase instead of merging to keep the history clear. Thanks.

**Pull request closed without merging**